core/src/main/java/com/microsoft/applicationinsights/internal/channel/common/TelemetryBuffer.java

```diff
@@ -220,12 +234,14 @@ public void flush() {
      *
      * @return The list of {@link Telemetry} instances that are ready to be sent
      */
-    private List<T> prepareTelemetriesForSend() {
-        ++generation;
+    private Collection<T> prepareTelemetriesForSend() {
+        generation.incrementAndGet();
 
-        final List<T> readyToBeSent = telemetries;
+        // Create an unmodifiable view of the collection so that the items cannot be edited
+        final Collection<T> readyToBeSent = new ArrayList<>(Collections.unmodifiableCollection(telemetries));
 
-        telemetries = new ArrayList<T>(maxTelemetriesInBatch);
+        // Clear telemetry for next batch
+        telemetries.clear();
```

Comment:
This seems to be broken, what if another thread pushed data to telemetries between L241 and L244?

Reply:
This might be a valid point. I will double check some references.

Reply:
I think you still want the reference swap, but use `AtomicReference`. Using `clear` could unintentionally remove some items